### PR TITLE
completed node and linked list

### PR DESCRIPTION
**Pull request by eobaah**

### CODE REVIEW

File: spec/linkedlist.js [hunk 1]
```diff
@@ -0,0 +1,339 @@
+import chai, { expect } from 'chai'
+import chaiChange from 'chai-change'
+import LinkedList from '../src/linkedlist'
+
+chai.use(chaiChange)
+
+describe('LinkedList', () => {
+  'use strict'
```
Comment (AbrahamFergie) [inline spec/linkedlist.js]: with ES6 syntax this string is no longer needed

File: spec/linkedlist.js [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+      myLinkedList.insert("Tesla")
+      expect(myLinkedList.head.data).to.equal("BMW")
+    })
+  })
+
+  context('insert()', () => {
+    it('Inserts multiple nodes (with the provided data) to an empty linked list, and checks the value of the head ', () => {
+      const myLinkedList = new LinkedList()
+      myLinkedList.insert("BMW")
+      myLinkedList.insert("Tesla")
```
Comment (AbrahamFergie) [inline spec/linkedlist.js]: so many of these tests have the same context, and the same 3 lines in them. To keep a good economic code file I would suggest just having all of your different expects in one context function, this way it is more readable for the reviewer and for you in the future.

File: spec/linkedlist.js [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+  })
+
+  context('find( input )', () => {
+    it('Returns -1 if node is not found', () => {
+      const myLinkedList = new LinkedList()
+      myLinkedList.insert("BMW")
+      myLinkedList.insert("Tesla")
+      myLinkedList.insert("Lexus")
+      myLinkedList.insertFirst("Mercedes")
+      expect(myLinkedList.find("Ford")).to.equal(-1)
```
Comment (AbrahamFergie) [inline spec/linkedlist.js]: you have 6 different input function tests separated out and a lot of repeating code:
context('contains(input)', () => {
    it('Determines whether or not the list contains the provided data', () => {
      const myLinkedList = new LinkedList()
      myLinkedList.insert("BMW")
      myLinkedList.insert("Tesla")
      myLinkedList.insert("Lexus")
      myLinkedList.insertFirst("Mercedes")
      expect(myLinkedList.contains("Lexus")).to.equal(true)
      expect(myLinkedList.contains("BMW")).to.equal(true)
      expect(myLinkedList.contains("Lexus")).to.equal(true)
      expect(myLinkedList.contains("Ford")).to.equal(false)
      expect(myLinkedList.find("Ford").data).to.equal("Ford")
      expect(myLinkedList.find("Ford")).to.equal(-1)
    })
  })
looks like the first and third test here is a duplicate but you can see that all these expects will still work and you will save a lot of space in your file by making the more concise and easier to read.

File: spec/node.js [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import chai, { expect } from 'chai'
+import chaiChange from 'chai-change'
+import Node from '../src/node'
+
+chai.use(chaiChange)
+
+describe('Node', () => {
+  'use strict'
```
Comment (AbrahamFergie) [inline spec/node.js]: same suggestion as the previous comment.

File: src/linkedlist.js [hunk 5]
```diff
@@ -0,0 +1,219 @@
+'use strict'
```
Comment (AbrahamFergie) [inline src/linkedlist.js:1]: just a reminder

File: src/linkedlist.js [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+    this.next = undefined;
+  }
+}
+
+export default class LinkedList {
+  constructor() {
+    this.head = undefined
+    this.tail = undefined
+    this.size = 0
+
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: white space here

File: src/linkedlist.js [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+      }
+      if(currentNode === this.tail && input !== this.tail.data){
+        return -1
+      } else {
+        currentNode = currentNode.next
+      }
+    }
+  }
+
+  insertBefore(before, after) {
```
Comment (AbrahamFergie) [inline src/linkedlist.js:74]: indentation here should be re-adjusted

File: src/linkedlist.js [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+      currentNode.next = beforeNode.next
+      beforeNode.next = currentNode
+      return currentNode
+    }
+    currentNode = currentNode.next
+  }
+  return after + " does not exist in this linked list"
+}
+
+  insertAfter(before, after) {
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: indentation re-adjustment

File: src/linkedlist.js [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+  let afterNode = new Node(after)
+  let currentNode = this.head
+
+
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: white space

File: src/linkedlist.js [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+      return true
+    } else {
+      return false
+    }
+  }
+
+  remove() {
+      let currentNode = this.head
+      // if list is empty
+      if(this.size === 0) {
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: these indentations should be corrected

File: src/linkedlist.js [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+        return "There is nothing to remove"
+    }
+
+      // if list has one item
+      if (this.size === 1) {
+        this.head = this.tail = undefined
+        this.size--
+        return this
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: same as this line and the 3 above.

File: src/linkedlist.js [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+      while(currentNode.next.next !== this.tail) {
+        currentNode = currentNode.next
+      }
+      this.tail = currentNode.next
+      currentNode.next.next = undefined
+      this.size--
+      return this
+  }
+
+  removeFirst() {
```
Comment (AbrahamFergie) [inline src/linkedlist.js]: indentation correction needed here as well.

File: src/node.js [hunk 13]
```diff
@@ -1,31 +1,40 @@
...
-      return this.next
-    }
+  getNext() {
+    return this.next
   }
 
+  getPrev() {
+    return this.prev
+  }
+
```
Comment (AbrahamFergie) [inline src/node.js:31]: white space

**Review by AbrahamFergie — APPROVED** [review]
I really don't want to be so picky, I just really want to hold everyone to a high standard so that we get used to what will be required of us in professional standards. I've had a couple PR's like this too and it really did help me stay more cognisant.

File: spec/linkedlist.js (hunk 3 above)
Comment (eobaah) [reply spec/linkedlist.js]: Thanks. I didn't know this could be done. Awesome!

File: spec/node.js (hunk 4 above)
Comment (eobaah) [reply spec/node.js]: Good catch

**Pull request merged**